src/KubernetesClient/Extensions.cs

```diff
@@ -0,0 +1,214 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Reactive;
+using System.Reactive.Concurrency;
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using k8s.Controllers;
+using k8s.Informers;
+using k8s.Informers.Cache;
+using k8s.Informers.Notifications;
+using k8s.Models;
+using Microsoft.Extensions.Logging;
+using Microsoft.Rest.TransientFaultHandling;
+using YamlDotNet.Serialization.NodeTypeResolvers;
+
+namespace k8s
+{
+    public static class Extensions
+    {
+        /// <summary>
+        /// Removes an item from the dictionary
+        /// </summary>
+        /// <param name="source">The source dictionary</param>
+        /// <param name="key">The key for which item should be removed</param>
+        /// <param name="result">The value of the object that was removed, or <see langword="null"/> if value was not present in dictionary</param>
+        /// <typeparam name="TKey">The type of key</typeparam>
+        /// <typeparam name="TValue">The type of value</typeparam>
+        /// <returns><see langword="true"/> if the object was removed from dictionry, or <see langword="false"/> if the specific key was not present in dictionary</returns>
+        public static bool Remove<TKey,TValue>(this IDictionary<TKey,TValue> source, TKey key, out TValue result)
+        {
+            result = default;
+            if (!source.TryGetValue(key, out result))
+                return false;
+            source.Remove(key);
+            return true;
+        }
+        /// <summary>
+        /// Tries to remove item from the queue
+        /// </summary>
+        /// <param name="queue">Source queue</param>
+        /// <param name="result">The result if dequeue was successful, other <see langword="null"/></param>
+        /// <typeparam name="T">The type of items in queue</typeparam>
+        /// <returns><see langword="true"/> if dequeue was successful, otherwise <see langward="false"/></returns>
+        public static bool TryDequeue<T>(this Queue<T> queue, out T result)
+        {
+            result = default;
+            if (queue.Count == 0)
+                return false;
+            try
+            {
+                result = queue.Dequeue();
+            }
+            catch (InvalidOperationException)
```

Comment:
given that this only occurs if the queue is empty (and you test for that above) isn't this redundant?

Reply:
There is no lock so there's a chance that Count became zero when we actually execute Dequeue (which is why it's done in catch block). I'm avoiding expensive throw if queue is empty, but still protecting against race condition.

Reply:
`Queue<T>` is fundamentally not thread-safe: it must not be shared by multiple threads without proper serialization. If you have observed an `InvalidOperationException` being thrown, it would suggest improper sharing and/or locking.

Reply:
Removed catch block and made it internal.

Reply:
`TryPeek` has been updated, but the issue remains for `TryDequeue`.

Reply:
Should be fixed now